src/torcontrol.cpp

```diff
@@ -0,0 +1,580 @@
+    //  If there is not a whole line to read, evbuffer_readln returns NULL
+    while((line = evbuffer_readln(input, &n_read_out, EVBUFFER_EOL_CRLF)) != NULL)
+    {
+        std::string s(line, n_read_out);
+        free(line);
+        if (s.size() < 4) // Short line
+            continue;
+        // <status>(-|+| )<data><CRLF>
+        self->message.code = atoi(s.substr(0,3).c_str());
+        self->message.lines.push_back(s.substr(4));
+        char ch = s[3]; // '-','+' or ' '
+        if (ch == ' ') {
+            // Final line, dispatch reply and clean up
+            if (self->message.code >= 600) {
+                // Dispatch async notifications to async handler
+                // Synchronous and asynchronous messages are never interleaved
+                self->async_handler(*self, self->message);
+            } else {
+                if (!self->reply_handlers.empty()) {
+                    // Invoke reply handler with message
+                    self->reply_handlers.front()(*self, self->message);
+                    self->reply_handlers.pop_front();
+                } else {
+                    LogPrint("tor", "tor: Received unexpected sync reply %i\n", self->message.code);
+                }
+            }
+            self->message.Clear();
+        }
+    }
+}
+
+void TorControlConnection::eventcb(struct bufferevent *bev, short what, void *ctx)
+{
+    TorControlConnection *self = (TorControlConnection*)ctx;
+    if (what & BEV_EVENT_CONNECTED) {
+        LogPrint("tor", "tor: Succesfully connected!\n");
+        self->connected(*self);
+    } else if (what & (BEV_EVENT_EOF|BEV_EVENT_ERROR)) {
+        if (what & BEV_EVENT_ERROR)
+            LogPrint("tor", "tor: Error connecting to Tor control socket\n");
+        else
+            LogPrint("tor", "tor: End of stream\n");
+        self->Disconnect();
+        self->disconnected(*self);
+    }
+}
+
+bool TorControlConnection::Connect(const std::string &target, const ConnectionCB& connected, const ConnectionCB& disconnected)
+{
+    if (b_conn)
+        Disconnect();
+    // Parse target address:port
+    struct sockaddr_storage connect_to_addr;
+    int connect_to_addrlen = sizeof(connect_to_addr);
+    if (evutil_parse_sockaddr_port(target.c_str(),
+        (struct sockaddr*)&connect_to_addr, &connect_to_addrlen)<0) {
+        LogPrintf("tor: Error parsing socket address %s\n", target);
+        return false;
+    }
+
+    // Create a new socket, set up callbacks and enable notification bits
+    b_conn = bufferevent_socket_new(base, -1, BEV_OPT_CLOSE_ON_FREE);
+    if (!b_conn)
+        return false;
+    bufferevent_setcb(b_conn, TorControlConnection::readcb, NULL, TorControlConnection::eventcb, this);
+    bufferevent_enable(b_conn, EV_READ|EV_WRITE);
+    this->connected = connected;
+    this->disconnected = disconnected;
+
+    // Finally, connect to target
+    if (bufferevent_socket_connect(b_conn, (struct sockaddr*)&connect_to_addr, connect_to_addrlen) < 0) {
```

Comment:
Don't we want logic to retry this every n seconds if it fails? If I launch Bitcoin and Tor a minute later, I would expect that it would be detected and used after a few minutes.